proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,408 @@
+
+Worker processes exit unconditionally after completing work. The monitoring
+thread spawns a new process for any exited process to replenish the process
+pool. Exited processes include crashed processes.  Processes that crash are
+detected by the monitoring thread allowing error information to be communicated
+back through the system.
+
+The worker process (Clang.exe) will be updated to include additional command
+options that indicate it should be launched as a worker and establish the IPC
+mechanism with a monitoring thread owned by the Compiler Process Pool.
+
+## Architecture Diagram
+```mermaid
+flowchart TD
+    subgraph CompilerProcessPool["Compiler Process Pool"]
+      subgraph MonitorThread1["Monitor Thread (1)"]
+        subgraph JsonRPC ["IPC Dispatcher"]
+        end
+      end
+      subgraph MonitorThread2["Monitor Thread (2)"]
+        subgraph JsonRPC2 ["IPC Dispatcher"]
+        end
+      end
+    end
+
+    subgraph Process1["process 1"]
+        ClangProcess1["clang.exe (1)"]
+        db[("compiled<br/>shader")]
+    end
+
+    subgraph Process2["process 2"]
+        ClangProcess2["clang.exe (2)"]
+        db2[("compiled<br/>shader")]
+    end
+
+    subgraph HighLevelCApi["Compiler Api (High-Level)"]
+        HighLevelApiBehaviorLogic("implements sync/async policy, packs/unpacks parameters")
+    end
+
+    subgraph LowLevelCApi["Compiler Api (Low-Level)"]
+        LowLevelApi("communicates with process pool, manages inputs/results")
+    end
+
+    style Process1 stroke-dasharray:10,10
+    style Process2 stroke-dasharray:10,10
+  
+    ApiCallDispatcher{"Api Dispatcher<br/>(find available process and dispatch work)"}
+
+    Api["Compile/CompileAsync()"] --> HighLevelApiBehaviorLogic <--> LowLevelCApi <--> |"dispatch and monitor call"| ApiCallDispatcher
+
+    ApiCallDispatcher -->JsonRPC<--> | json-rpc<br/>protocol | ClangProcess1 --> db
+
+    ApiCallDispatcher --> JsonRPC2<-->| json-rpc<br/>protocol | ClangProcess2 --> db2
+```
+
+### Calling apis
+A high-level api call starts in the api entrypoint implementation which will
+call into low-level apis and implement sync/async behaviors. The low-level
+c-api is responsible for finding an available worker process (clang.exe) to
+perform the work.  If a worker is not available the high-level api behavior
+logic will queue the compilation request and block the entrypoint for a
+synchronous calls waiting for the next available worker to be freed up or
+for and Asynchronous call return immediately giving the caller enough context
+to monitor or cancel the compilation.
+
+The api implementations, dispatcher and process pool all live in-process to
+the calling application.
+
+The dispatching system uses IPC and a message protocol to communicate with
+each worker process.  Each worker process will know how to package up results
+and communicate them back over the IPC mechanism to their monitoring threads
+which will bubble up those results via the low-level apis.
```

Comment:
So all source files and generated binaries are transferred across this IPC channel?

Reply:
I was thinking that the caller would specify output paths as part of their command line args packed to the compile request.  So clang would just do what it does today with generated binaries etc.